### ISSUE TITLE
Import Cycle

### ISSUE THREAD

**Issue by ninnemana** [curt-labs/GoAPI#21]
The apicontext helper has introduced an import cycle:

models/customer -> models/product -> helpers/apicontext -> models/customer

In my opinion, the apicontext helper should be abstracted away from the customer object as much as possible. I think the simplest approach is to remove the DataContext completely and map the individual properties rather than a struct. 

I guess if this problem has been created solely for the purpose of mocking data, it's probably being done wrong.

**Comment by DavidVaini** [comment]
I think originally the DataContext was created to transport the important information to the models.(brandID, key, etc), but the customer and customerUser was only addded because we needed a way to mock and demock a datacontext. In order to generate an API key for testing(since now every function requires a api key), you also need to generate a Customer User and in order to generate a Customer User you need to generate a Customer. I decided to throw it in the DataContext because then in your tests you could mock one up, simply use it, and then demock it when you are done. If we got rid of Customer and CustomerUser that would be fine, but we still need a seperate helper then to mock up a customer and customer user. Unless you want to have that mock up code for a customer and customer user and api key to be repeated in every test.

**Comment by ninnemana** [comment]
if it's just for testing, don't use the customer package???

**Comment by ninnemana** [comment]
interface{} maybe?

**Comment by DavidVaini** [comment]
Well If you try to test with a fake Customer/CustomerUser/ApiKey/ApiKeyToBrand...etc then there arnt any records in the db, and then you cant test. For example if you are testing a GetLatestParts, in order to get any records, you need a record in the api key table, apiKeyToBrand table. If there are no records there, you wont get any parts because of the joins.

**Comment by ninnemana** [comment]
so why not create a test helper that queries for that information at random?

**Comment by ninnemana** [comment]
you're tests will be pointless if they bomb on import cycles, right?

**Comment by DavidVaini** [comment]
exactly, didnt notice the import cycles with Customer Tests when I first created this.

Thats why I think it would be a good idea to have a separate helper for mocking a Customer, Customer User, APIKey, and APIKeyToBrand records. Then for each test, you simply import that helper and use it.

The reason I originally put it in the DataContext is because it felt like I was mockiing up all the things I would need in order to use the Data Context appropriately.

**Comment by ninnemana** [comment]
otherwise maybe the DataContext helper just can't have references to other packages. Other packages would only reference the DataContext. Rather than storing CustomerUser, store CustomerUser.Id; rather than Customer; store Customer.Id

**Comment by DavidVaini** [comment]
yeah but then if you dont have other references to other packages like customer, then you cant use the customers methods for creating customers and creating customer users and creating api keys, etc. And if you cant create any of those things you cant mock up the API Key and brand.

If you have two seperate mocking helpers. One that mocks up a Customer>CustomerUser>APIKey>APIKeyToBrand>Brand, etc (That whole chain)

Then you have the DataContext Mock which really is only a struct that stores, everything it currently does minus the Customer and CustomerUser (cause its not really needed).

**Comment by DavidVaini** [comment]
Then inside your test, you call the Customer.Mock() and then after you get your information, then you fill out your DataContext Struct, then you call Customer.DeMock() to remove the customer after testing.

**Comment by ninnemana** [comment]
Are you saying the DataContext helper needs the ability to call Customers.Add?

**Comment by DavidVaini** [comment]
So I fixed one of the import cycles:
https://github.com/curt-labs/GoAPI/commit/5c8bb2d4e18d084a10db9e66a757fa03a0ad6100

However I did notice a seperate one. In the Customer Tests (both model test and controller tests) they have imports to products. Products have an import to Customer. Inside these customer tests its testing cart integration stuff which should be done inside the products test and not the customer test, so these need to be moved.

Products need to have an import to Customers, so customer stuff can not and should not import products (even in tests - they share the same package).
